### ISSUE TITLE
Redirect should encode URLs if it needs to

### ISSUE THREAD

**Issue by miloush** [aspnet/AspNetCore#4919]
All these issues are result of people trying to redirect to a URL that is not all ASCII. Resolution seems to be "you should have encoded your URL".

https://github.com/aspnet/Home/issues/2678
https://github.com/aspnet/Mvc/issues/7701
https://github.com/aspnet/KestrelHttpServer/issues/2508
https://github.com/aspnet/KestrelHttpServer/issues/2296
https://github.com/aspnet/Security/issues/1646
https://github.com/aspnet/HttpAbstractions/issues/971
https://github.com/aspnet/Routing/issues/513
https://github.com/aspnet/Mvc/issues/7529
https://github.com/aspnet/HttpAbstractions/issues/907
https://github.com/aspnet/Mvc/issues/6609
https://github.com/aspnet/Home/issues/2106
https://github.com/madskristensen/Miniblog.Core/issues/87

However, the fact that the RedirectResult puts the URL into a HTTP header (which is the ultimate source of the issue) is an implementation detail that the caller cannot magically assume. It is therefore responsibility of e.g. RedirectResultExecutor to meet the requirements and ensure the URL is properly formatted in case it decides to use HTTP headers for redirection.

**Comment by sksk571** [comment]
Any update on this issue? We are migrating .NET framework app and this looks like a regression compared to .NET framework. Documentation for Redirect doesn't state that the URL should be properly encoded, therefore it's a responsibility of Redirect to encode it before passing into header.

Looking at the .NET4.8 source code, it appears that Response.Redirect selectively encodes whitespace and non-ASCII characters in URL by calling UrlEncodeRedirect: https://referencesource.microsoft.com/#System.Web/HttpResponse.cs,fb8b07aa51863d27

**Comment by rjgotten** [comment]
I fail to see how applying a percent-encoding on characters that are not ASCII-safe can lead to any kind of double-encoding bug.

Characters that have syntactic value in URLs are all ASCII-safe and would not be touched.
Percent-encoded sequences are entirely composed of ASCII-safe cahracters and would not be touched.

It is literally impossible to hit a double-encoding problem.

**Comment by mirhagk** [comment]
Yeah there definitely seems to be a lot of confusion in this discussion. The escaping being discussed is not a general URI escaping but just a non-ASCII specific escaping. There is no chance of a double-encode as a properly encoded string would not contain any non-ASCII characters. The only "breaking change" is that RedirectResult actually works now instead of throwing exceptions, and the chances that someone is depending on the framework throwing an exception here seems pretty low.

It's very counter-intuitive, especially as IRIs gain more popularity and browsers, javascript and the old .NET framework all support them fine without prior encoding. It seems odd that .NET core would take such a big step back in terms of internationalization.

**Comment by Basssiiie** [comment]
@rjgotten Nice solution! It also already seems to be used in the ASP.NET Core code base in other places. 😄 

- [UriHelper.cs](https://source.dot.net/#Microsoft.AspNetCore.Http.Extensions/UriHelper.cs,174)
- [CreatedResult.cs](https://source.dot.net/#Microsoft.AspNetCore.Mvc.Core/CreatedResult.cs,62)
- [AcceptedResult.cs](https://source.dot.net/#Microsoft.AspNetCore.Mvc.Core/AcceptedResult.cs,57)

And a bunch of others. Using it in to `RedirectResult` as well would bring `RedirectResult` in line with the others.
